python_modules/dagster/dagster/_core/host_representation/external_data.py

```diff
@@ -1238,6 +1242,32 @@ def __new__(
         auto_observe_interval_minutes: Optional[float] = None,
         owners: Optional[Sequence[str]] = None,
     ):
+        metadata = normalize_metadata(check.opt_mapping_param(metadata, "metadata", key_type=str))
+
+        # backcompat logic for execution type specified via metadata
+        if SYSTEM_METADATA_KEY_ASSET_EXECUTION_TYPE in metadata:
+            val = metadata[SYSTEM_METADATA_KEY_ASSET_EXECUTION_TYPE]
+            if not isinstance(val, TextMetadataValue):
+                check.failed(
+                    f"Expected metadata value for key {SYSTEM_METADATA_KEY_ASSET_EXECUTION_TYPE} to be a TextMetadataValue, got {val}"
+                )
+            metadata_execution_type = AssetExecutionType.str_to_enum(val.value)
+            if execution_type is not None:
+                check.invariant(
+                    execution_type == metadata_execution_type,
+                    f"Execution type {execution_type} in metadata does not match type inferred from metadata {metadata_execution_type}",
+                )
+            execution_type = metadata_execution_type
+        else:
+            execution_type = (
+                check.opt_inst_param(
+                    execution_type,
+                    "execution_type",
+                    AssetExecutionType,
+                )
+                or AssetExecutionType.MATERIALIZATION
+            )
+
```

Comment:
Even though we are sticking with metadata for now to specify `AssetExecutionType` on  `AssetsDefinition`, We still need this check because when an `ExternalAssetNode` represents a source asset there is no such metadata.